### DISCUSSION TITLE
Testing and Work Prior to Beta

### DISCUSSION THREAD

**Discussion by Corionis** [GrokSoft/ELS#158] [General 1. Discussions]
Closing-in on a full Beta Release. However there are things to be done before that.

- Still testing various scenarios and need to perform a full regression test on all platforms
  - There are so many permutations in the ways ELS may be used testing is a challenge
  - The large refactors are hopefully over. A consequence of the length of development, the complexity and the maturing of the code over time. Several areas have been almost completely rewritten.
    - In other words "_What was I thinking when I wrote that??_" :laughing: 
- Update internal documentation
- Update GitHub Wiki documentation

The Beta Release will update the face of the GitHub site from Release 3.1 to Beta 4.0.

Beta is very important. _THAT_ is when I must ask everyone to use it, work it, and tell me what's wrong or needs to be changed. 

After almost 3 years of work and a 6x expansion of the code it's coming together (finally :grin:)

**Comment by Corionis** [comment]
While implementing more-complex use cases have uncovered a handful of GUI bugs related to paths in Jobs.

Working through those but also have fall home projects in the works. Might take a little while.

**Comment by Corionis** [comment]
Having an established test environment loads all that existing data. However, when starting from a newly-installed ELS getting everything working the first time has exposed a few rough spots and bugs. Making changes in a variety of places to make starting-up a new ELS environment easier.

**Comment by Corionis** [comment]
Along the way am fixing some logical problems that stem from the evolution of the code over 3 years, again. For instance removing the local or remote publish distinction from the Operations Tool so that, when put in a Job Task, the Subscriber connection defines whether it is a local or remote operation. Other similar issues have been fixed. Details when this batch of changes is complete. Soon, 1-2 days.

**Comment by Corionis** [comment]
Took a little longer than anticipated, migraine issues. The **14 October 2024** build is a significant clean-up and polish.

  * Added the ability to select a system directory for Libraries - Locations, Sources, Multi-source, and the temporary directory.
  * Adjusted Job Task Origins publisher/subscriber/hint server display and select list handling.
  * Removed the distinction between Local and Remote Publish from the Operations tool. That is now determined by the Subscriber connection.
  * <i>Important:</i> The Operations value for Local and Remote Publish has changed. Previous Publisher values are invalid and <b>will cause an exception</b>.
  * Subscriber listener (--remote S) operations no longer require a Publisher if an Authentication Keys file is specified.
  * Subscriber operations do not require a Hint Server and will be ignored. Hints are handled by Publishers.
  * Removed Hint Tracker/Server combobox and file value from Operations so the Hint Tracker/Sever Origin is defined in a Job Task.
  * <i>Important:</i> To enable the use of Hints in a Job Task define Hint Keys in the Operations Tool.
  * Fixed updating Jobs when a Library is renamed or its key is changed, or if Hint Keys are removed.
  * Fixed dynamic updating of the Browser and Hints dialog as Hints are added, run, etc.
  * Fixed restoring mnemonics, arrows, etc. on Cancel in Settings.
  * Changed the default Browser tab location from bottom to top and fixed a related bug.

**Comment by Corionis** [comment]
The large changes are settling down. Now it's testing and specific fixes and changes. 

Will attempt to be more granular with Issues and Changelist notes to provide more detailed problem and change information.

**Comment by Corionis** [comment]
Have been deploying a pre-Beta implementation in a real-world 3-remote node backup arrangement.

This will test the various ELS mechanisms in a "production" environment interactively and automation using Linux scripting and ELS Jobs running on a cron schedule.

Version 4.1 will add the Tools and extensions needed to fully automate backup and other processes entirely inside ELS.  
See: [Version 4.1 Release ](https://github.com/Corionis/ELS/milestone/2).

**Comment by Corionis** [comment]
Moving ahead with more testing and progress on #113 Update documentation. Finally approaching the Beta release of 4.0.0. A lot of reading and typing over the coming days. This includes internal documentation, GitHub documentation, and code comments and headers.

**Comment by Corionis** [comment]
Although I thought the big changes were done Issue #180 "_Changing current working directory fails_" is taking a lot of time and a lot of path-related changes. This involves work to support:

- Absolute paths
- Relative paths
- Symbolic links on Linux-based platforms
- Starting ELS in a directory other than its configuration directory (-C option)
- Libraries relative to the configuration directory, as physical directories or symbolic links, or located elsewhere

When this work is completed ELS will be able to be executed from anywhere with any configuration with or without using the included launchers and fully support symbolic links.

Lots of testing and cross-platform process testing is required.

**Note**: ELS supports so many permutations of configuration, options and use patterns it is difficult for one person (me) to think of and test all of it in a comprehensive fashion. I really need folks to beat on it and communicate their findings and suggestions in these Discussions and creating Issues as appropriate. A GitHub account is required to collaborate.

**Comment by Corionis** [comment]
This has been rather painful. Many permutations and places to change, and sometimes changed again. To be honest am suffering a little burn-out on these issues so progress has been slow. So much to test but slogging through it. When finished several bugs will be resolved and ELS will be more flexible.

**Comment by Corionis** [comment]
And ... now I'm down sick :nauseated_face:

**Comment by Corionis** [comment]
31 January 2025
Finally getting back to it.  This is an interim code commit with no new executable deployment pending a lot of testing.

- Changed the ELS-Navigator.sh script for Linux to allow arguments. If none the last configuration is used.
- Fixed a save bug in the Sleep Tool.
- Fixed a busy cursor bug on a remote action warning.
- Fixed #179.
- Fixed #180.
- Added support for symbolic links in the -C configuration directory (~/.els).
- Fixed #183.
- Fixed #184.

**Comment by Corionis** [comment]
Continuing a long test sequence, fixing bugs and tweaking things along the way.

Today:
 * Fixed a bug related to restarting & reopening a remote subscriber
 * Fixed a minor Job app header bug
 * Added a "Generate/Run" button when selecting a Job directly from the Jobs menu

Have been a bit distracted. Still need to test cross-platform from Windows and macOS related to the #180 _Changing current working directory fails_ low-level I/O changes.

That is why there has not been an update to the distribution files yet.

**Comment by Corionis** [comment]
Linux-Windows and Windows-Linux seem to be working with the new low-level changes, symbolic links, etc.

Will be moving to macOS next with Linux, then against Windows.

Bug reports and comments encouraged.

**Comment by Corionis** [comment]
Found bugs #185 and #186 while working in Windows. 

Windows shortcut creation now checks for the 260 character limit, shows a warning and will not create an invalid shortcut.  If a command line is too long create an Operation and put it in a Job. Job command lines are much shorter because the Job and Tasks contain all the options.

Execution of Navigator in a Job now works correctly.

1. If Navigator is executed as the first step in Job any subsequent tasks in that Job will be execute within Navigator.
2. If Navigator is a later step it is run in order. If it is a foreground Job then Navigator is executed within the Logger, i.e. closing the Logger will close Navigator.

The startup dialog should now disappear correctly.

**Reply by Corionis** [reply]
btw:These changes have not been tested on Linux or Mac yet. Tomorrow.

**Comment by Corionis** [comment]
Proceeding with special testing between Linux, Windows and macOS as peers. Pounding on the changes related to #180 _Changing current working directory fails_, support for symbolic links and the low-level I/O changes.

**Comment by Corionis** [comment]
Finally deployed a new set of executables after all the low-level changes and cross-platform testing on all three platforms.

Moving ahead.

**Comment by Corionis** [comment]
Beginning the final work before Beta.

Want to support headless and non-graphical systems #149, then do a lot of typing to complete internationalization and update all the internal and GitHub Wiki documentation.

See [Version 4.0 Release](https://github.com/Corionis/ELS/milestone/1) Milestone for what's left. At this moment (1 May 2025) the Issues of the milestones are not in the planned order. A minor GitHub bug being worked on.

**Comment by Corionis** [comment]
Just an update. Currently working on https://github.com/Corionis/ELS/issues/65 Complete internationalization. The original code up to Version 4.0 was not i18n so there is a lot to go through. Trying to be comprehensive and not break it 😁

That said the locale resource file will need optimization are some point to refactor duplicate and near-duplicate phrases.

Want all the major code changes in the coming, ever so slowly, Beta Release.

**Comment by Corionis** [comment]
The deployment yesterday is code complete for Beta. Finally. Finished internationalization and fixed a bug when saving Libraries that could cause Libraries to become unresponsive.

Honestly the language file needs to be optimized and shrunk. It's kinda messy. That will be done during Beta.

Have tested it on all three platforms as best as one person can. Am currently working on the internal and Wiki documentation. Then a final round of tests and the Beta release.

**Comment by Corionis** [comment]
Deployed and merged Beta 4.0 but can't fully test updates until it's committed.

Found a bug - still testing the Beta release full-cycle.

**Comment by Corionis** [comment]
There is a bug in updating macOS in the previous deployment. The only way to cure the problem is to download the macOS tar and overwrite any existing installation. Apologies. Some things cannot be tested - like updates - until the next deployment.

**Comment by Corionis** [comment]
**ELS Beta 4.0 Release**

After 3 years and 8 months creating 9 projects to put together ELS on Linux, Mac and Windows it's ready for critical eyes.

There are some rough spots - like no Bookmark editor. But it seems to work in as many ways as one person can think of and test. The real issue is ELS can work with so many different operations and options there are likely bugs here and there.

All the internal Help documentation, GitHub [Wiki](https://github.com/Corionis/ELS/wiki) documentation, and the [user web site](https://corionis.github.io/ELS/) have been updated. 

The Version-4.0.0 branch has been merged to master. Further Beta work will continue in the Version-4.0.0 branch and merged to master when a body of work is completed and ready for deployment. When work begins on [Version 4.1](https://github.com/Corionis/ELS/milestone/2) a new branch will be created from the master branch.

Any feedback in [Discussions](https://github.com/Corionis/ELS/discussions) or [Issues](https://github.com/Corionis/ELS/issues) would be _greatly appreciated_. 

Now is the time to starting talking to me and each other.
